### PR TITLE
Android add discover

### PR DESCRIPTION
**Pull request by joonhaengHeo**
#### Problem
What is being fixed?  
* Android Test Application is not supported dns discover. (Python chip-ctrl-tool command is "discover -all")

#### Change overview
 - Add NsdManagerDiscoverService class.
 - Support discover function in AddressCommissioning activiry.

#### Testing
How was this tested? (at least one bullet point required)
* Check to be able to discover in Android App.

### CODE REVIEW

**Review by bzbarsky-apple — APPROVED** [review]

**Review by austinh0 — CHANGES_REQUESTED** [review]
Similar to what Damian said - if you are going to add `NsdManagerServiceDiscover` in the platform code, then you should also implement the platform [DnssdImpl::ChipDnssdBrowse](https://github.com/project-chip/connectedhomeip/blob/6e22d13c0ffa3e640d19b39a7755df8c1942b77e/src/platform/android/DnssdImpl.cpp#L78).

I can accept doing mDNS discovery purely in the app code (i.e. use `NsdManager` directly in the app), at least as a temporary solution, but we should not mix these two approaches. 

Will review the PR in more detail after this is resolved.

**Comment by joonhaengHeo** [comment]
@austinh0 Does it mean that it should be implemented in a jni format similar to the implementation of Service Resolver?

**Comment by austinh0** [comment]
> @austinh0 Does it mean that it should be implemented in a jni format similar to the implementation of Service Resolver?

Ideally, yes. Service Resolver is used like so: 
`CHIPDeviceController::UpdateDevice` -> common native code -> `platform/android/DnssdImpl::ChipDnssdResolve` -> (Java) `ServiceResolver.resolve` -> (Java) `ChipMdnsCallback.handleServiceResolve` -> `AndroidChipPlatform-JNI::handleServiceResolve` -> `DnssdImpl::HandleResolve`

So I would expect mDNS browse to follow a similar path starting with `CHIPDeviceController:DiscoverCommissionableNodes`, with some Java `ServiceBrowser.browse`.

This is a fair amount of effort since you need to write Java -> C++ (and vice versa) translation for the `DnssdService` struct, which is why I mentioned I'm also ok with a simple app-only implementation that directly uses NsdManager as a temporary solution.

**Review by austinh0 — CHANGES_REQUESTED** [review]
Looks good overall.

Is there a reason we call `discoverCommissionableNodes`, wait, and the check `getDiscoveredDevice` instead of having some sort of callback method that is invoked when a device is discovered? I haven't looked at the native implementation in a while.

File: src/platform/android/java/chip/platform/ChipMdnsCallbackImpl.java [hunk 1]
```diff
@@ -17,12 +17,27 @@
...
       long callbackHandle,
       long contextHandle);
+
+  public String[] getAttributeKeys(Map<String, byte[]> attributes) {
+    return attributes.keySet().toArray(new String[attributes.size()]);
+  }
+
+  public byte[] getAttributeData(Map<String, byte[]> attributes, String key) {
+    return attributes.get(key);
+  }
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/ChipMdnsCallbackImpl.java:42]: OK. I understand.

File: src/platform/android/java/chip/platform/AndroidChipPlatform.java [hunk 2]
```diff
@@ -17,19 +17,32 @@
...
       ConfigurationManager cfg,
       ServiceResolver resolver,
+      ServiceBrowse browse,
       ChipMdnsCallback chipMdnsCallback) {
     setBLEManager(ble);
     setKeyValueStoreManager(kvm);
     setConfigurationManager(cfg);
-    setServiceResolver(resolver, chipMdnsCallback);
+    setServiceResolver(resolver, browse, chipMdnsCallback);
   }
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/AndroidChipPlatform.java]: OK. I understand.

File: src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt [hunk 3]
```diff
@@ -46,10 +58,42 @@ class AddressCommissioningFragment : Fragment() {
...
+      updateSpinnerHandler.sendEmptyMessageDelayed(0, 5000)
+      discoverBtn.isEnabled = false;
+    }
+  }
+
+  private fun updateSpinner() {
+    val deviceController = ChipClient.getDeviceController(requireContext())
+    for(i: Int in 0..10) {
+      val device = deviceController.getDiscoveredDevice(i) ?: break
+      ipAddressList.add("${device.ipAddress},${device.discriminator}")
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt]: resolved.

File: src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt [hunk 4]
```diff
@@ -46,10 +58,42 @@ class AddressCommissioningFragment : Fragment() {
...
+      val deviceController = ChipClient.getDeviceController(requireContext())
+      deviceController.discoverCommissionableNodes()
+      updateSpinnerHandler.sendEmptyMessageDelayed(0, 5000)
+      discoverBtn.isEnabled = false;
+    }
+  }
+
+  private fun updateSpinner() {
+    val deviceController = ChipClient.getDeviceController(requireContext())
+    for(i: Int in 0..10) {
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt]: Resolved.
I confirmed that it is 10 in the code below.

https://github.com/project-chip/connectedhomeip/blob/master/src/controller/CHIPCommissionableNodeController.h#L69

https://github.com/project-chip/connectedhomeip/blob/8e62dd6794d86f73c7b3639c6b7633708b9530c5/src/include/platform/CHIPDeviceConfig.h#L1098

File: src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt [hunk 5]
```diff
@@ -46,10 +58,42 @@ class AddressCommissioningFragment : Fragment() {
...
+        ArrayAdapter(requireContext(), android.R.layout.simple_spinner_dropdown_item, ipAddressList)
+      discoverListSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
+        override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
+          val address = ipAddressList[position].split(",")[0].trim()
+          val discriminator = ipAddressList[position].split(",")[1].trim()
+          addressEditText.setText(address)
+          discriminatorEditText.setText(discriminator)
+        }
+
+        override fun onNothingSelected(parent: AdapterView<*>?) {}
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt]: Resolved.

File: src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+import android.net.nsd.NsdServiceInfo;
+import android.net.wifi.WifiManager;
+import android.net.wifi.WifiManager.MulticastLock;
+import android.os.Handler;
+import android.os.Message;
+import android.util.Log;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public class NsdManagerServiceBrowse implements ServiceBrowse {
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java]: OK. I understand.

File: src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt [hunk 7]
```diff
@@ -46,10 +58,42 @@ class AddressCommissioningFragment : Fragment() {
         )
       )
     }
+
+    discoverBtn.setOnClickListener { _ ->
+      val deviceController = ChipClient.getDeviceController(requireContext())
+      deviceController.discoverCommissionableNodes()
+      updateSpinnerHandler.sendEmptyMessageDelayed(0, 5000)
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt]: Resolved.

File: src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt [hunk 8]
```diff
@@ -46,10 +58,42 @@ class AddressCommissioningFragment : Fragment() {
...
+    val deviceController = ChipClient.getDeviceController(requireContext())
+    for(i: Int in 0..10) {
+      val device = deviceController.getDiscoveredDevice(i) ?: break
+      ipAddressList.add("${device.ipAddress},${device.discriminator}")
+    }
+    requireActivity().runOnUiThread {
+      discoverListSpinner.adapter =
+        ArrayAdapter(requireContext(), android.R.layout.simple_spinner_dropdown_item, ipAddressList)
+      discoverListSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
+        override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt]: Resolved.

File: src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt [hunk 9]
```diff
@@ -46,10 +58,42 @@ class AddressCommissioningFragment : Fragment() {
...
+
+        override fun onNothingSelected(parent: AdapterView<*>?) {}
+      }
+    }
   }
 
   companion object {
     private const val TAG = "AddressCommissioningFragment"
+    private const val SERVICE_TYPE = "_matterc._udp"
+    private val ipAddressList = ArrayList<String>()
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/AddressCommissioningFragment.kt]: I agree that it doesn't have to be static.
The reason for not using DiscoveredDevice is that when creating a Spinner, if it is not a String, it is inconvenient to define a new class for spinner. Therefore, it is implemented using String.

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 10]
```diff
@@ -351,6 +351,55 @@ JNI_METHOD(void, updateDevice)(JNIEnv * env, jobject self, jlong handle, jlong f
...
+
+    AndroidDeviceControllerWrapper * wrapper = AndroidDeviceControllerWrapper::FromJNIHandle(handle);
+    const Dnssd::DiscoveredNodeData * data   = wrapper->Controller()->GetDiscoveredDevice(idx);
+
+    if (data == nullptr)
+    {
+        return nullptr;
+    }
+
+    jclass discoveredDevicecls = env->FindClass("chip/devicecontroller/ChipDeviceController$DiscoveredDevice");
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/controller/java/CHIPDeviceController-JNI.cpp]: Resolved.

File: src/controller/java/src/chip/devicecontroller/ChipDeviceController.java [hunk 11]
```diff
@@ -196,6 +196,19 @@ public void updateDevice(long fabricId, long deviceId) {
...
   }
 
+  public void discoverCommissionableNodes() {
+    discoverCommissionableNodes(deviceControllerPtr);
+  }
+
+  public class DiscoveredDevice {
+    public long discriminator;
+    public String ipAddress;
+  }
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/controller/java/src/chip/devicecontroller/ChipDeviceController.java]: Resolved.

File: src/controller/java/src/chip/devicecontroller/ChipDeviceController.java [hunk 12]
```diff
@@ -196,6 +196,19 @@ public void updateDevice(long fabricId, long deviceId) {
     updateDevice(deviceControllerPtr, fabricId, deviceId);
   }
 
+  public void discoverCommissionableNodes() {
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/controller/java/src/chip/devicecontroller/ChipDeviceController.java:203]: Resolved.

File: src/platform/android/java/chip/platform/AndroidChipPlatform.java [hunk 13]
```diff
@@ -73,12 +86,13 @@ public native void handleUnsubscribeComplete(
   private native void setConfigurationManager(ConfigurationManager manager);
 
   // for ServiceResolver
-  private void setServiceResolver(ServiceResolver resolver, ChipMdnsCallback chipMdnsCallback) {
+  private void setServiceResolver(
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/AndroidChipPlatform.java]: Resolved.

File: src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+  private static final String TAG = NsdManagerServiceBrowse.class.getSimpleName();
+  private static final long DISCOVER_SERVICE_TIMEOUT = 3000;
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java]: Resolved. I will change to 5 sec. And, I also change to wait application delay to 6 sec.

AddressCommissioningFragment.kt
     scope.launch {
        delay(**6000**)
        updateSpinner()
        discoverBtn.isEnabled = true
      }

[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java]: After testing, delay value is changed to 7 sec.

File: src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+      final String serviceType,
+      final long callbackHandle,
+      final long contextHandle,
+      final ChipMdnsCallback chipMdnsCallback) {
+    startDiscover(serviceType, callbackHandle, contextHandle, chipMdnsCallback);
+
+    Message msg = new Message();
+    msg.what = 0;
+    msg.obj = callbackHandle;
+    handler.sendMessageDelayed(msg, DISCOVER_SERVICE_TIMEOUT);
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java]: Resolved.

File: src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public class NsdManagerServiceBrowse implements ServiceBrowse {
+  private static final String TAG = NsdManagerServiceBrowse.class.getSimpleName();
+  private static final long DISCOVER_SERVICE_TIMEOUT = 3000;
+  private final NsdManager nsdManager;
+  private MulticastLock multicastLock;
+
+  private HashMap<Long, NsdManagerDiscovery> callbackMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java]: Resolved.

File: src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+    public void onStopDiscoveryFailed(String serviceType, int errorCode) {
+      Log.w(TAG, "Failed to stop discovery service '" + serviceType + "': " + errorCode);
+    }
+
+    @Override
+    public void onDiscoveryStopped(String serviceType) {
+      Log.w(TAG, "Succeed to stop discovery service '" + serviceType);
+    }
+
+    public void sendCallback() {
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdManagerServiceBrowse.java]: Resolved.

File: src/platform/android/java/chip/platform/NsdManagerServiceResolver.java [hunk 18]
```diff
@@ -81,8 +81,6 @@ public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
             Log.w(
                 TAG,
                 "Failed to resolve service '" + serviceInfo.getServiceName() + "': " + errorCode);
-            chipMdnsCallback.handleServiceResolve(
-                instanceName, serviceType, null, 0, callbackHandle, contextHandle);
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/NsdManagerServiceResolver.java]: I have restored it.
I was mistaken for an implementation before.

File: src/platform/android/java/chip/platform/ServiceBrowse.java [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+package chip.platform;
+
+/** Interface for resolving network services. */
+public interface ServiceBrowse {
+  /** TO-DO */
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/ServiceBrowse.java]: Updated.

File: src/platform/android/java/chip/platform/ServiceBrowse.java [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+  public void browse(
+      final String serviceType,
+      final long callbackHandle,
+      final long contextHandle,
+      final ChipMdnsCallback chipMdnsCallback);
+}
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/platform/android/java/chip/platform/ServiceBrowse.java]: Resolved.

**Review by austinh0 — APPROVED** [review]

**Review by jmeg-sfy — APPROVED** [review]

**Pull request merged**